Implement the only modifier.

lijunle opened the pull request:
**DO NOT MERGE**

Resolve #132 

I hack the test count number to show the *right* number and passed cases number. From my understand, such information should be pass through the promise chain.

Besides, `ava.serial.only` has no support in this PR yet.

/cc @vdemedes @sindresorhus

---

Screenshot:

![screenshot_2015-11-12_23-48-19](https://cloud.githubusercontent.com/assets/1296500/11122835/e9d793d4-8997-11e5-86d0-bff5c3cc1374.png)

lijunle commented on lib/runner.js:
```diff
@@ -203,16 +209,21 @@ Runner.prototype.run = function () {
 		.catch(noop)
 		.then(function () {
+			// HACK the test count number if only modifier is shown
+			stats.testCount = tests.only.length ? tests.only.length : stats.testCount;
```
The variable `testCount` is maintain by the `addTest`, `addOnlyTest`, `addSerialTest` function, which I think that is not suitable. A better way is to return such information (`testCount`, `failCount`) from the promise, then calculate the final stats from at the end of the chain.

However, it involves more code change if doing so. Any ideas?

vdemedes replied:
I think it's perfectly fine, I wouldn't call it a hack.

sindresorhus commented:
Awesome @lijunle :) Both @vdemedes and I are traveling, so we probably won't be able to review this until monday.

// @jamestalmage

lijunle commented:
@sindresorhus That is OK, please keep in mind that, there is a hack. I am not going to merge until resolve the hack with a good way (unless a direction).

vdemedes commented on lib/runner.js:
```diff
@@ -203,16 +209,21 @@ Runner.prototype.run = function () {
 		})
 		.then(function () {
-			return self.serial(tests.serial);
+			return tests.only.length ? self.concurrent(tests.only) : [];
```
it can be just `return self.concurrent(tests.only)`, since `tests.only` defaults to `[]`.

vdemedes commented:
Everything looks good so far, I pointed out some things in the diff. One question I have in mind, should `only` tests be executed serially on in parallel? Btw, mocha allows only one `only` test, so maybe we should also do that? And it would solve the serial vs parallel question too.

lijunle commented:
> One question I have in mind, should only tests be executed serially on in parallel?

My question is *no*. If the guy want to do so, he should do `ava.serial.only`. However such API is not implemented in this RP.

> Btw, mocha allows only one only test, so maybe we should also do that?

Mocha allow several `only` modifiers. For example, if there are 10 test case, and 3 of them are marked as `only`, those 3 test cases will be run.

vdemedes commented:
> Mocha allow several only modifiers. For example, if there are 10 test cases, and 3 of them are marked as only, those 3 test cases will be run.

I guess I was running some out-dated version then.

vdemedes commented:
@lijunle Could you please rebase your PR (to pick up latest fixes for windows tests) and address comments in the diff? Happy to merge this ASAP.

lijunle commented:
OK, do you want to modify the comment about the *hack*?

vdemedes commented:
Yeah, the comment is not really needed and this - https://github.com/sindresorhus/ava/pull/204/files#r45035490

lijunle commented:
Rebased, let us wait for the CI. :smile_cat: 

---

However, although I removed the comment, I still consider that is a *hack*.

The stats should be collected from `_runTestWithHooks` function and pass back to the promise chain in the `Runner.prototype.run` function. At the end of the promise chain, calculate the final stats from the input.

Pseudo code like this:

```js
Runner.prototype.run = function () {
	var self = this;
	return eachSeries(tests.before, this._runTest.bind(this)) // resolve with { pass: 1, failed: 2, skipped: 3 }
		.then(function (stats) {
			return Object.assign(stats, self.serial(tests.serial)); // i.e. { pass: 2, failed: 3, skipped: 4 }
		})
		.then(function (stats) {
			return Object.assign(stats, self.concurrent(tests.concurrent)); // object with { pass, failed, skipped }
		})
		.then(function (stats) {
			return Object.assign(stats, eachSeries(tests.after, self._runTest.bind(self))); // object with { pass, failed, skipped }
		})
		.then(function (stats) {
			self.stats.passCount = stats.pass; // calculate from the input of promise chain.
			return stats; // external check could depends on return value of run function, more stable. 
		});
};
```

vdemedes commented:
The refactoring and cleanup work is ahead, so don't worry about this 1 line ;)

lijunle commented:
I understand. It is up to you to design the pattern, that is just my suggestion.

jamestalmage commented:
@lijunle 

I was actually taking a stab at this earlier when I got pulled away to some other stuff.
Since you are pretty far down the road, I am holding off.

I just squashed my work and pushed it up: https://github.com/jamestalmage/ava/commit/4568347c6ae230e8f03bb20f93f5430ce7bfff58

Use it if it helps.

My other thoughts.

1. I would **really** like to avoid the Mocha behavior of only allowing a single test in exclusive mode. I want to be able to put `only` on however many tests I chose and have them be the ones that run. This is especially important since we don't have grouping yet.

2. I like your idea of `test.serial.only`, etc.

3. I think creating stats at the beginning is making all this hard. Just put flags in each individual test and use `this.tests.concurrent.filter` to create the stats and get at the tests you want (see the diff I linked).

Good work! And thanks for the help.

vdemedes commented:
@lijunle Thank you!

sindresorhus commented:
@jamestalmage Any improvements welcome in a follow-up pull request ;)

jamestalmage commented:
Yeah, sorry! 
I've got a lot going on today and am lagging way behind the conversation.
